### ISSUE TITLE
make registries admin-configurable

### ISSUE THREAD

**Issue by krancour** [akuity/kargo#1139]
Supersedes #1106

Follow-up to #1058

Most of the time, details for how to talk to a given image registry can be inferred from an image's "prefix," but not always:

* Sometimes the API address is different from the hostname.

    Example: Images with no prefix _or_ with hostname `docker.io` in their prefix (e.g. `nginx` or `docker.io/nginx`) have an API address https://registry-1.docker.io

* Sometimes a registry has a "default namespace."

    Example: Images with no prefix or only a hostname in their prefix (e.g. `nginx` or `docker.io/nginx`) need to be prefixed with `library/` when interacting with the API. e.g. `nginx` becomes `library/nginx`.

* Rate limits cannot be inferred.

For well known, public registries like Docker Hub or Quay, it's sensible to hard-code accommodations for such idiosyncrasies, but when such idiosyncrasies arise in self-hosted registries, we need some mechanism for users to define the rules themselves.

I recommend that registries be configured _globally_ by a Kargo admin and _not_ be configured at the project level. #1058 already treats each registry configuration as global singletons so that:

* The registry cache is shared across projects

* Many projects all talking to the same registry are rate-limited as a pool instead of individually.

Ultimately, I believe this boils down to permitting registry idiosyncrasies to be defined by a Kargo admin either/both:

1. In a `values.yaml` file used during Kargo installation/upgrade. (The chart would incorporate this into a `ConfigMap`.)

1. A manually maintained `ConfigMap`.

**Comment by jessesuen** [comment]
> I recommend that registries be configured globally by a Kargo admin and not be configured at the project level. https://github.com/akuity/kargo/pull/1058 already treats each registry configuration as global singletons so that:

I don't think we should switch to this model. Global configuration of registries would not work in a harder tenancy environment where different sets of credentials may be needed to access different image repositories. This would be similar to how git repositories need to have different credentials.

For example, within github, you might have `ghcr.io/example/repo1` and `ghcr.io/example/repo2`, for two different teams, both under the `example` org, but each with different credentials.

Now, this shouldn't stop people from using our global credential feature to use a single set of credentials for their entire org. I agree this should be an option available to users. But the granular level of image regiestries is still needed for harder tenancy.

**Comment by krancour** [comment]
@jessesuen and I spoke offline to clear up that this is isn't about credentials, but about smoothing over registry idiosynchrasies like we do here:

https://github.com/akuity/kargo/blob/8534a0c3d83e1eb729c2cd596b8feb81a4c1b608/internal/image/registry.go#L29-L50

We've agreed that if idiosyncrasies with popular public registries become known, we can amend the above.

If someone self-hosting a registry runs into such a problem, we can re-prioritize this issue at that time.

**Comment by krancour** [comment]
While I expected this to be an issue, it has yet to surface in the wild, so I am going to close this and we can re-consider it if/when we learn of self-hosted registries Kargo doesn't play nice with.

**Comment by bbigras** [comment]
Does kargo work with harbor? #1106

**Comment by krancour** [comment]
@bbigras thanks for finding that. Unknown. I will re-open this pending investigation.

**Comment by bbigras** [comment]
Possible still important.

**Comment by Kavinraja-G** [comment]
Bumping... Looking for Harbor support 👀

**Comment by krancour** [comment]
This has come up again fairly recently. It's not clear yet that it's something we need to act on, but it's worth keeping this open for now.

**Comment by krancour** [comment]
Related to #5571 

This issue speaks to the more nuanced approach that #5571 suggests may eventually be necessary.
